pixelsmith.py

```diff
@@ -0,0 +1,193 @@
+from time import sleep
+from typing import Callable
+import torch
+from tqdm.auto import tqdm
+
+from comfy.model_base import SDXL
+import comfy.model_management
+import comfy.samplers
+import comfy.sample
+import comfy.utils
+import latent_preview
+from comfy_extras.nodes_custom_sampler import Noise_RandomNoise
+from nodes import MAX_RESOLUTION
+from .utils import (
+    ModelSamplingType,
+    PixelsmithSampler,
+    apply_chess_mask,
+    combine_hires_guided,
+    create_gradient_border,
+    crop,
+    crop_assign,
+    get_value_coordinates,
+    max_denoise,
+)
+
+
+# NOTE: Based on SamplerCustomAdvanced
+#
+# Pixelsmith is based on https://github.com/Thanos-DB/Pixelsmith/blob/f15d725d19a782ae340c7c1db6cd301550fcf6cf/pixelsmith_pipeline.py#L1715
+# licensed under GPL v3
+class Pixelsmith:
+    @classmethod
+    def INPUT_TYPES(s):
+        return {
+            "required": {
+                "noise": ("NOISE",),
+                "guider": ("GUIDER",),
+                "sampler": ("SAMPLER",),
+                "sigmas": ("SIGMAS",),
+                "latent_image": ("LATENT",),
+                "slider": ("FLOAT", {"default": 0.5, "min": 0, "max": 1.0, "step": 0.01}),
+                "patch_size": ("INT", {"default": 1024, "min": 16, "max": MAX_RESOLUTION, "step": 8}),
+            }
+        }
+
+    RETURN_TYPES = ("LATENT", "LATENT")
+    RETURN_NAMES = ("output", "denoised_output")
+    FUNCTION = "sample"
+    CATEGORY = "sampling/custom_sampling"
+
+    def sample(
+        self,
+        noise: Noise_RandomNoise,
+        guider: comfy.samplers.CFGGuider,
+        sampler: comfy.samplers.KSAMPLER,
+        sigmas: torch.Tensor,
+        latent_image,
+        slider: float,
+        patch_size: int,
+    ):
+        model: SDXL = guider.model_patcher.model
+        ms: ModelSamplingType = guider.model_patcher.model.model_sampling
+        p_sampler = PixelsmithSampler(sampler)
+        slider_step = int((len(sigmas) - 1) * slider)
+
+        patch_size = patch_size // 8
+
+        latent = latent_image
+        latent_image = latent["samples"]
+        latent = latent.copy()
+        latent_image = comfy.sample.fix_empty_latent_channels(guider.model_patcher, latent_image)
+        latent["samples"] = latent_image
+
+        noise_mask = None
+        if "noise_mask" in latent:
+            noise_mask = latent["noise_mask"]
+
+        x0_output = {}
+        callback = latent_preview.prepare_callback(guider.model_patcher, sigmas.shape[-1] - 1, x0_output)
+
+        disable_pbar = not comfy.utils.PROGRESS_BAR_ENABLED
+
+        # Pixelsmith
+        guided_noise = noise.generate_noise(latent)
+        guided_at = lambda step: ms.noise_scaling(sigmas[step], guided_noise, latent_image, max_denoise(ms, sigmas))

```

Comment:
After stepping through the diffusers code, it looks like this actually should be something like:
```suggestion
        guided_at = lambda step: AddNoise().add_noise(guider.model_patcher, noise, sigmas[step].unsqueeze(0), latent)[0]['samples']
```

(will require an import of `from comfy_extras.nodes_custom_sampler import AddNoise`)

Reply:
I may be wrong on this, as this is what AddNoise is basically doing anyway, except it additionally scales the output... https://github.com/comfyanonymous/ComfyUI/blob/ef85058e977f886c88d4a30b819708b1168f39a4/comfy_extras/nodes_custom_sampler.py#L700